akka-stream-tests/src/test/scala/akka/stream/io/TlsSpec.scala

```diff
@@ -0,0 +1,312 @@
+    }
+
+    object EmptyBytesFirst extends PayloadScenario {
+      def inputs = List(ByteString.empty, ByteString("hello")).map(SendBytes)
+      def output = ByteString("hello")
+    }
+
+    object EmptyBytesInTheMiddle extends PayloadScenario {
+      def inputs = List(ByteString("hello"), ByteString.empty, ByteString(" world")).map(SendBytes)
+      def output = ByteString("hello world")
+    }
+
+    object EmptyBytesLast extends PayloadScenario {
+      def inputs = List(ByteString("hello"), ByteString.empty).map(SendBytes)
+      def output = ByteString("hello")
+    }
+
+    object ImpatientRHS extends PayloadScenario {
+      override def flow =
+        Flow[SslTlsInbound]
+          .mapConcat {
+            case SessionTruncated       ⇒ SessionTruncated :: Nil
+            case SessionBytes(s, bytes) ⇒ bytes.map(b ⇒ SessionBytes(s, ByteString(b)))
+          }
+          .take(5)
+          .mapAsync(10, x ⇒ later(500.millis, system.scheduler)(Future.successful(x)))
+          .via(super.flow)
+      override def rightClosing = IgnoreCancel
+
+      val str = "abcdef" * 100
+      def inputs = str.map(send)
+      def output = ByteString(str.take(5))
+    }
+
+    object SessionRenegotiationBySender extends PayloadScenario {
+      def inputs = List(send("hello"), NegotiateNewSession.default, send("world"))
+      def output = ByteString("helloNEWSESSIONworld")
+    }
+
+    // difference is that the RHS engine will now receive the handshake while trying to send
+    object SessionRenegotiationByReceiver extends PayloadScenario {
+      val str = "abcdef" * 100
+      def inputs = str.map(send) ++ Seq(NegotiateNewSession.default) ++ "hello world".map(send)
+      def output = ByteString(str + "NEWSESSIONhello world")
+    }
+
+    val logCipherSuite = Flow[SslTlsInbound]
+      .map {
+        var session: SSLSession = null
+        def setSession(s: SSLSession) = {
+          session = s
+          system.log.debug(s"new session: $session (${session.getId mkString ","})")
+        }
+
+        {
+          case SessionTruncated ⇒ SendBytes(ByteString("TRUNCATED"))
+          case SessionBytes(s, b) if s != session ⇒
+            setSession(s)
+            SendBytes(ByteString(s.getCipherSuite) ++ b)
+          case SessionBytes(s, b) ⇒ SendBytes(b)
+        }
+      }
+
+    object SessionRenegotiationFirstOne extends PayloadScenario {
+      override def flow = logCipherSuite
+      def inputs = NegotiateNewSession.withCipherSuites("TLS_RSA_WITH_AES_128_CBC_SHA") :: send("hello") :: Nil
+      def output = ByteString("TLS_RSA_WITH_AES_128_CBC_SHAhello")
+    }
+
+    object SessionRenegotiationFirstTwo extends PayloadScenario {
+      override def flow = logCipherSuite
+      def inputs = NegotiateNewSession.withCipherSuites("TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA") :: send("hello") :: Nil
+      def output = ByteString("TLS_ECDHE_RSA_WITH_AES_128_CBC_SHAhello")
+    }
+
+    val scenarios =
```

Comment:
Really nice! I had a much more limited solution, nice to see it generalized to perfection!